File: backend/src/main/java/com/orderhub/catalog/service/ProductService.java
package com.orderhub.catalog.service;

import com.orderhub.catalog.dto.ProductRequest;
import com.orderhub.catalog.dto.ProductResponse;
import com.orderhub.catalog.dto.ProductStatusRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;
import java.util.UUID;

/**
 * Service interface for managing product catalog operations.
 * <p>
 * Provides methods for browsing active products, viewing product details,
 * and admin operations for creating, updating, and managing product status.
 * </p>
 *
 * @since 1.0.0
 */
public interface ProductService {

    /**
     * Retrieves a paginated list of active products with optional filters.
     * <p>
     * Supports filtering by name (case-insensitive substring match) and price
     * range.
     * All filter parameters are optional.
     * </p>
     *
     * @param name     optional name filter
     * @param minPrice optional minimum price filter (inclusive)
     * @param maxPrice optional maximum price filter (inclusive)
     * @param pageable pagination and sorting parameters
     * @return page of active products matching the filter criteria
     */
    Page<ProductResponse> getActiveProducts(String name, BigDecimal minPrice,
            BigDecimal maxPrice, Pageable pageable);

    /**
     * Retrieves a single product by ID.
     *
     * @param id the product ID
     * @return the product details
     * @throws com.orderhub.common.exception.ResourceNotFoundException if product
     *                                                                 not found
     */
    ProductResponse getProductById(UUID id);

    /**
     * Creates a new product.
     * <p>
     * Also creates an associated inventory record with quantity 0.
     * </p>
     *
     * @param request the product creation request
     * @return the created product
     * @throws com.orderhub.common.exception.DuplicateResourceException if SKU
     *                                                                  already
     *                                                                  exists
     */
    ProductResponse createProduct(ProductRequest request);

    /**
     * Updates an existing product.
     * <p>
     * Validates that SKU is unique (excluding the product being updated).
     * </p>
     *
     * @param id      the product ID
     * @param request the product update request
     * @return the updated product
     * @throws com.orderhub.common.exception.ResourceNotFoundException  if product
     *                                                                  not found
     * @throws com.orderhub.common.exception.DuplicateResourceException if SKU
     *                                                                  already
     *                                                                  exists
     */
    ProductResponse updateProduct(UUID id, ProductRequest request);

    /**
     * Updates a product's active status.
     * <p>
     * Used to activate or deactivate products. Deactivated products
     * will not appear in public product listings.
     * </p>
     *
     * @param id      the product ID
     * @param request the status update request
     * @return the updated product
     * @throws com.orderhub.common.exception.ResourceNotFoundException if product
     *                                                                 not found
     */
    ProductResponse updateProductStatus(UUID id, ProductStatusRequest request);
}

File: frontend/src/app/core/services/product.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Product, ProductPage } from '../models/product.model';

/**
 * Service for managing product operations
 */
@Injectable({
  providedIn: 'root',
})
export class ProductService {
  private readonly http = inject(HttpClient);
  private readonly API_URL = '/api/v1/products';

  /**
   * Retrieves a paginated list of products with optional filters
   */
  getProducts(params: {
    page?: number;
    size?: number;
    name?: string;
    minPrice?: number;
    maxPrice?: number;
    sort?: string;
  }): Observable<ProductPage> {
    let httpParams = new HttpParams();

    if (params.page !== undefined && params.page !== null) {
      httpParams = httpParams.set('page', params.page.toString());
    }
    if (params.size !== undefined && params.size !== null) {
      httpParams = httpParams.set('size', params.size.toString());
    }
    if (params.name) {
      httpParams = httpParams.set('name', params.name);
    }
    if (params.minPrice !== undefined && params.minPrice !== null) {
      httpParams = httpParams.set('minPrice', params.minPrice.toString());
    }
    if (params.maxPrice !== undefined && params.maxPrice !== null) {
      httpParams = httpParams.set('maxPrice', params.maxPrice.toString());
    }
    if (params.sort) {
      httpParams = httpParams.set('sort', params.sort);
    }

    return this.http.get<ProductPage>(this.API_URL, { params: httpParams });
  }

  /**
   * Retrieves a single product by ID
   */
  getProductById(id: string): Observable<Product> {
    return this.http.get<Product>(`${this.API_URL}/${id}`);
  }
}
